core/src/main/java/io/grpc/internal/AbstractServerStream.java

```diff
@@ -31,256 +31,223 @@
+        listener().halfClosed();
       }
-      writeStatusToTrailers(status);
-      sendStreamAbortToClient(status, stashedTrailers);
     }
-  }
 
-  @Override
-  public boolean isClosed() {
-    return super.isClosed() || listenerClosed;
-  }
+    @Override
+    protected ServerStreamListener listener() {
+      return listener;
+    }
 
-  /**
-   * Notifies the remote client that this stream has aborted.
-   */
-  protected abstract void sendStreamAbortToClient(Status status, Metadata trailers);
+    /**
+     * Called in the network thread to process the content of an inbound DATA frame from the client.
+     *
+     * @param frame the inbound HTTP/2 DATA frame. If this buffer is not used immediately, it must
+     *              be retained.
+     * @param endOfStream {@code true} if no more data will be received on the stream.
+     */
+    public void inboundDataReceived(ReadableBuffer frame, boolean endOfStream) {
+      if (inboundHalfClosed) {
+        frame.close();
+        return;
+      }
+      // Deframe the message. If a failure occurs, deframeFailed will be called.
+      deframe(frame, endOfStream);
+    }
 
-  /**
-   * Fires a half-closed event to the listener and frees inbound resources.
-   */
-  private void halfCloseListener() {
-    if (inboundPhase(Phase.STATUS) != Phase.STATUS && !listenerClosed) {
-      closeDeframer();
-      listener().halfClosed();
+    /**
+     * Notifies failure to the listener of the stream. The transport is responsible for notifying
+     * the client of the failure independent of this method.
+     *
+     * <p>Unlike {@link #close(Status, Metadata)}, this method is only called from the
+     * transport. The transport should use this method instead of {@code close(Status)} for internal
+     * errors to prevent exposing unexpected states and exceptions to the application.
+     *
+     * @param status the error status. Must not be {@link Status#OK}.
+     */
+    public final void transportReportStatus(Status status) {
+      Preconditions.checkArgument(!status.isOk(), "status must not be OK");
+      closeListener(status);
     }
-  }
 
-  /**
-   * Closes the listener if not previously closed and frees resources.
-   */
-  private void closeListener(Status newStatus) {
-    if (!listenerClosed) {
-      listenerClosed = true;
-      closeDeframer();
-      listener().closed(newStatus);
+    /**
+     * Indicates the stream is considered completely closed and there is no further opportunity for
+     * error. It calls the listener's {@code closed()} if it was not already done by {@link
+     * #transportReportStatus}.
+     */
+    public void complete() {
+      closeListener(Status.OK);
     }
-  }
 
-  @Override public Attributes attributes() {
-    return Attributes.EMPTY;
+    /**
+     * Closes the listener if not previously closed and frees resources.
+     */
+    private void closeListener(Status newStatus) {
+      if (!listenerClosed) {
+        listenerClosed = true;
```

Comment:
@ejona86 I think we also need to set `inboundHalfClosed = true`, as we are shutting down the deframer. This then fixes #1795, as in `inboundDataReceived()` we are only checking for `inboundHalfClosed`.